File: recursion/string_palindrome.js
/**
 * @param {string} str
 * @param {number} start
 * @param {number} end
 * @return {boolean}
 */
function isPalindrome(str, start = 0, end = str.length - 1) {
	if (start >= end)
		return true;
	return str.charAt(start) == str.charAt(end)
		&& isPalindrome(str, start + 1, end - 1);
}

/**
 * @param {string} str
 */
function run(str) {
	console.log(`${str}\t${isPalindrome(str) ? "true" : "false"}`);
}

console.log("word\tisPalindrome");
run("gabriel");
run("abba");
run("bella");
run("001100");
run("");

File: recursion/string_palindrome.c
#include <stdio.h>
#include <string.h>

int is_palindrome(char *str, int start, int end) {
	if (start >= end)
		return 1;
	return str[start] == str[end] && is_palindrome(str, start + 1, end - 1);
}

void run(char *str) {
	printf("%s\t%d\n", str, is_palindrome(str, 0, strlen(str) - 1));
}

int main() {
	printf("\nword\ts_palindrome\n");

	run("gabriel");
	run("abba");
	run("bella");
	run("00100");
	run("");

	printf("\n");
	return (0);
}
